internal/state/graph/httproute.go

```diff
@@ -1,146 +1,359 @@
 package graph
 
 import (
+	"fmt"
+
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/gateway-api/apis/v1beta1"
 
 	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/conditions"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/validation"
 )
 
+// Rule represents a rule of an HTTPRoute.
+type Rule struct {
+	BackendGroup BackendGroup
+	// ValidMatches indicates whether the matches of the rule are valid.
+	// If the matches are invalid, NGK should not generate any configuration for the rule.
+	ValidMatches bool
+	// ValidFilters indicates whether the filters of the rule are valid.
+	// If the filters are invalid, the data-plane should return 500 error provided that the matches are valid.
+	ValidFilters bool
+}
+
+// ParentRef describes a reference to a parent in an HTTPRoute.
+type ParentRef struct {
+	// Gateway is the NamespacedName of the referenced Gateway
+	Gateway types.NamespacedName
+	// Idx is the index of the reference in the HTTPRoute.
+	Idx int
+}
+
 // Route represents an HTTPRoute.
 type Route struct {
 	// Source is the source resource of the Route.
 	// FIXME(pleshakov)
 	// For now, we assume that the source is only HTTPRoute.
 	// Later we can support more types - TLSRoute, TCPRoute and UDPRoute.
 	Source *v1beta1.HTTPRoute
+	// SectionNameRefs is a map of section names to the referenced NKG Gateways
+	SectionNameRefs map[string]ParentRef
+	// UnboundSectionNameRefs is a subset of SectionNameRefs that includes sections that could not be bound
+	// to the referenced Gateway. For example, because section does not exist in the Gateway.
+	UnboundSectionNameRefs map[string]conditions.Condition
+	// Conditions include Conditions for the HTTPRoute.
+	Conditions []conditions.Condition
+	// Rules include Rules for the HTTPRoute. Each Rule[i] corresponds to the ith HTTPRouteRule.
+	// If the Route is invalid, the field is nil
+	Rules []Rule
+	// Valid tells if the Route is valid.
+	// If it is invalid, NGK should not generate any configuration for it.
+	Valid bool
+}
+
+// GetAllBackendGroups returns BackendGroups for all rules with valid matches and filters in the Route.
+func (r *Route) GetAllBackendGroups() []BackendGroup {
```

Comment:
I'm struggling with a good name here.

> GetBackendGroupsForValidRules()

I don't want to use "ValidRules" because this makes me want to add `Valid()` method to the Rule:
```
func (r Rule) Valid() bool {
	return r.ValidMatches && r.ValidFilters	
}
```
However, if the rule is invalid because of ` r.ValidFilters`, we still need to add configuration for it in the data plane. So I can't use something like below in the code:
```
if !r.Valid() {
  // ignore (do not generate any config)
```

> GetValidBackendGroups()

this one doesn't really work because we don't really have invalid backend groups.

Reply:
What about `GetBackendGroupsForConfigurableRules()`? Downside is that it's super long haha

Reply:
when `r.ValidFilters` is false, if r.ValidMatches is true, the rule is still configurable - the data plane needs to handle a request to it, but return 500  :(

Reply:
I think refactoring BackendGroup (so that nginx/config package doesn't depend on graph package) might help arrive with a better abstraction to how represent backends for rules. So we might arrive at that name after refactoring

Reply:
Ok. Let's hold off then. Can you add a FIXME so we don't forget?

Reply:
https://github.com/nginxinc/nginx-kubernetes-gateway/pull/455/commits/3b5a532eb0495ab54a38e48c4a5acf7d73e8291f